composer/lib/dependabot/composer/metadata_finder.rb

```diff
@@ -28,27 +28,26 @@ def look_up_source_from_packagist
         return nil if packagist_listing&.fetch("packages", nil) == []
         return nil unless packagist_listing&.dig("packages", dependency.name.downcase)
 
-        version_listings =
-          packagist_listing["packages"][dependency.name.downcase].
-          select { |version, _| Composer::Version.correct?(version) }.
-          sort_by { |version, _| Composer::Version.new(version) }.
-          map { |_, listing| listing }.
-          reverse
-
-        potential_source_urls =
-          version_listings.
-          flat_map { |info| [info["homepage"], info.dig("source", "url")] }.
-          compact
-
-        source_url = potential_source_urls.find { |url| Source.from_url(url) }
-
-        Source.from_url(source_url)
+        version_listings = packagist_listing["packages"][dependency.name.downcase]
+        # Packagist returns an array of version listings sorted newest to oldest.
+        # So iterate until we find the first URL that appears to be a source URL.
+        #
+        # NOTE: Each listing may not have all fields because they are minified to remove duplicate elements:
+        # * https://github.com/composer/composer/blob/main/UPGRADE-2.0.md#for-composer-repository-implementors
+        # * https://github.com/composer/metadata-minifier
+        version_listings.each do |i|
+          [i["homepage"], i.dig("source", "url")].each do |url|
+            source_url = Source.from_url(url)
+            return source_url unless source_url.nil?
+          end
+        end
+        return nil # rubocop:disable Style/RedundantReturn
```

Comment:
Like `rubocop`, I too was surprised I needed to explicitly return `nil` here... but if I didn't, then tests started failing and when I dug in it was because it was returning one of the previous function calls...  `version_listings` IIRC.

Reply:
Oh, I think this is because in the last line of a method, `return <something>` is redundant. You can just write `<something>`. Because Ruby by default uses the result of the last line executed as the return value.

Reply:
oh, good call, I didn't even think about writing a plain `nil` here as I was only considering omitting it entirely... and indeed `nil` passes both CI and rubocop. 👍